runtime/common/src/asset_registry.rs

```diff
@@ -70,4 +74,50 @@ impl<
 		let metadata = OrmlAssetRegistry::metadata_by_location(&location)?;
 		Some(metadata.additional.fee_per_second)
 	}
-}
\ No newline at end of file
+}
+
+// Define a convertor to convert between a CurrencyId and the dia oracle keys using the metadata
+// stored in the asset registry
+pub struct AssetRegistryToDiaOracleKeyConvertor<Runtime>(PhantomData<Runtime>);
+
+impl<
+		Runtime: orml_asset_registry::Config<AssetId = CurrencyId, CustomMetadata = CustomMetadata>,
+	> Convert<Key, Option<(Vec<u8>, Vec<u8>)>> for AssetRegistryToDiaOracleKeyConvertor<Runtime>
+{
+	fn convert(spacewalk_oracle_key: Key) -> Option<(Vec<u8>, Vec<u8>)> {
+		let currency_id = match spacewalk_oracle_key {
+			Key::ExchangeRate(currency_id) => currency_id,
+		};
+
+		// Try to find the dia keys in the asset registry metadata
+		orml_asset_registry::Pallet::<Runtime>::metadata(currency_id).and_then(|metadata| {
+			let dia_keys = metadata.additional.dia_keys;
+			if dia_keys.blockchain.is_empty() || dia_keys.symbol.is_empty() {
+				return None;
+			}
+			return Some((dia_keys.blockchain, dia_keys.symbol));
+		});
+
+		// We didn't find the dia keys in the asset registry metadata
+		None
+	}
+}
+
+impl<
+		Runtime: orml_asset_registry::Config<AssetId = CurrencyId, CustomMetadata = CustomMetadata>,
+	> Convert<(Vec<u8>, Vec<u8>), Option<Key>> for AssetRegistryToDiaOracleKeyConvertor<Runtime>
+{
+	fn convert(dia_oracle_key: (Vec<u8>, Vec<u8>)) -> Option<Key> {
```

Comment:
Is it possible to have tuples that contain empty `Vec<u8>` here? In that case, should we check for empty vectors so we spare some storage iterations?

Reply:
Good point. I added an additional check, just in case.